pkg/agent/controller/networkpolicy/l7engine/l7visbility_watcher.go

```diff
@@ -0,0 +1,397 @@
+	trafficControlApi := tcController.CrdClient.CrdV1alpha2().TrafficControls()
+	_, err := trafficControlApi.Create(context.TODO(), tc, metav1.CreateOptions{})
+	if err != nil && !strings.Contains(err.Error(), "already exists") {
+		klog.ErrorS(err, "Failed to create TrafficControl")
+		return err
+	}
+	return nil
+}
+
+func (l7w *l7Watcher) postTrafficControl(tcName string, direction v1alpha2.Direction, podLabels, namespaceLabels map[string]string) error {
+	targetPort := &v1alpha2.OVSInternalPort{Name: config.L7NetworkPolicyTargetPortName}
+	var err error
+	if len(podLabels) != 0 {
+		err = createTrafficControl(tcName, namespaceLabels, podLabels, direction, v1alpha2.ActionMirror, targetPort, l7w.tcController)
+	} else {
+		err = createTrafficControl(tcName, namespaceLabels, nil, direction, v1alpha2.ActionMirror, targetPort, l7w.tcController)
+	}
+	if err != nil {
+		klog.ErrorS(err, "Error Creating Traffic Control")
+		return err
+	}
+	err = l7w.tcController.SyncTrafficControl(tcName)
+	if err != nil {
+		klog.ErrorS(err, "Error Syncing Traffic Control")
+		return err
+	}
+	return nil
+}
+
+func (l7w *l7Watcher) deleteTrafficControl(tcName string) error {
+	return l7w.tcController.CrdClient.CrdV1alpha2().TrafficControls().Delete(context.TODO(), tcName, metav1.DeleteOptions{})
+}
+
+func generateTCName(direction v1alpha2.Direction, podLabels, namespaceLabels map[string]string) (string, error) {
+	directionStr := "Both"
+	switch direction {
+	case v1alpha2.DirectionIngress:
+		directionStr = "Ingress"
+	case v1alpha2.DirectionEgress:
+		directionStr = "Egress"
+	case v1alpha2.DirectionBoth:
+		directionStr = "Both"
+	}
+	if podLabels == nil {
+		podLabels = make(map[string]string)
+	}
+	data := map[string]interface{}{
+		"direction":       directionStr,
+		"podLabels":       podLabels,
+		"namespaceLabels": namespaceLabels,
+	}
+	dataJSON, err := json.Marshal(data)
+	if err != nil {
+		klog.ErrorS(err, "Error creating hash for Traffic Control")
+		return "", err
+	}
+	hashval := sha256.Sum256(dataJSON)
+	hash := fmt.Sprintf("%x", hashval[:14])
+	return hash, nil
+}
+
+func (l7w *l7Watcher) addToLabelsToTCMap(labelsKey, tcName string) {
+	if tcArray, ok := l7w.labelsToTCMap[labelsKey]; ok {
+		l7w.labelsToTCMap[labelsKey] = append(tcArray, tcName)
+	} else {
+		l7w.labelsToTCMap[labelsKey] = []string{tcName}
+	}
+}
+
+func createLabelsKey(podLabels, namespaceLabels map[string]string) (string, error) {
+	data := map[string]interface{}{
+		"podLabels":       podLabels,
+		"namespaceLabels": namespaceLabels,
+	}
+	dataJSON, err := json.Marshal(data)
+	if err != nil {
+		klog.ErrorS(err, "Error creating hash for Traffic Control")
+		return "", err
+	}
+	hashval := sha256.Sum256(dataJSON)
+	key := fmt.Sprintf("%x", hashval[:14])
+	return key, nil
+}
+
+func (l7w *l7Watcher) startSuricataAndCreateTC(direction v1alpha2.Direction, nodeName string, podLabels, namespaceLabels map[string]string) error {
```

Comment:
when I think about the feature works, I don't think there is a need to create a TrafficControl CR. This is the current flow (based on my understanding):
- watch Pods for which l7 visibility is enabled
- create the appropriate TrafficControl CR
- the Node will observe the new TrafficControl CR and configure the datapath appropriately

throughout these steps, there is only one Node and one Agent involved. There is no need to create a CR and persist it in K8s. We can re-use the work that we did to implement the TrafficControl CRD, but we do not need to create an actual CR. That may require some refactoring in `pkg/agent/controller/trafficcontrol`.

That will also resolve the RBAC issue I pointed out in the manifest.

cc @tnqn for opinion

Reply:
Yes, the suggestion makes sense to me. There are also problems by using the TrafficControl that it cannot select the Pods with the annotation in some case.